File: src/cpp_basic_nodes/src/launch1_publisher.cpp
/*
 * This is a simple publisher that is a part of `launch1`
 * 
 * This node publishes a string message on a topic. 
 * It takes the following parameters
 * - l1pub_PubFreq [double]: The publishing frequency (in Hz) for the node
 * 
 * The node publishes to the following topics
 * - ~/pub_topic [std_msgs/String]: Publishing topic
 */

// Include basic header files
#include <ros/ros.h>
// Message header file for publisher
#include <std_msgs/String.h>

using namespace std;

int main(int argc, char** argv) {
    // Initialize the node
    ros::init(argc, argv, "launch1_publisher");
    // Create a node handler
    ros::NodeHandle nh("~");    // "~" -> Local namespace
    
    // Set the value of publishing frequency
    string param_name = "l1pub_PubFreq";
    double pub_freq = 0.1;  // Default publishing frequency
    if (nh.hasParam(param_name)) {  // If the parameter exists, try allotting value
        if (nh.getParam(param_name, pub_freq)) {
            ROS_INFO_STREAM("Publishing frequency set to " << 
                            pub_freq << " Hz from parameter '" << param_name << "'");
        } else {
            ROS_WARN_STREAM("Parameter '" << param_name << "' could not be parsed as double");
        }
    } else {
        ROS_INFO_STREAM("Could not find the parameter '" << param_name << "', using default value of " 
                        << pub_freq << " Hz instead");
    }
    
    // Set the rate of publisher
    ros::Rate rate_handler(pub_freq);
    
    // Create a publisher object
    ros::Publisher pub_obj = nh.advertise<std_msgs::String>("pub_topic", 100);
    
    // Message object
    double seq_num = 1;
    std_msgs::String msg_obj;
    // Main publishing loop
    while (ros::ok()) {
        // Create message
        msg_obj.data = to_string(seq_num);
        pub_obj.publish(msg_obj);
        seq_num += 1;
        ROS_DEBUG_STREAM("Published message [topic: " << pub_obj.getTopic() << "] '" << msg_obj.data << "'");
        ROS_INFO_STREAM("Published " << msg_obj.data);
        // Wait for the duration
        rate_handler.sleep();
    }
    
    return 0;
}

File: src/py_basic_nodes/scripts/launch1_publisher.py
#!/usr/bin/env python3

"""
    This is a simple publisher that is a part of `launch1`

    This node publishes a string message on a topic.
    It takes the following parameters
    - ~/l1pub_PubFreq [double]: The publishing frequency (in Hz) for the node

    The node publishes to the following topics
    - pub_topic [std_msgs/String]: Publishing topic
"""

# Import basic library
import rospy
import sys
# Import the modules for messages
from std_msgs.msg import String


# Main function
def main():
    # Initialize node
    rospy.init_node("launch1_publisher", argv=sys.argv)

    # Parameter for publishing frequency
    param_name = "{node_name}/l1pub_PubFreq".format(node_name=rospy.get_name()) # Need parameter in local scope
    pub_freq = 0.1  # Default publishing frequency
    try:
        pub_freq = float(rospy.get_param(param_name))
        rospy.loginfo("Publishing frequency set to {freq} Hz from parameter '{pname}'".format(
            freq=pub_freq, pname=param_name
        ))
    except ValueError as e1:
        rospy.logwarn("Parameter '{pname}' could not be parsed as float, value remains {val}: {err}".format(
            pname=param_name, val=pub_freq, err=e1
        ))
    except KeyError as e2:
        rospy.loginfo("Could not find the parameter '{pname}', using default value of {val}: {err}".format(
            pname=param_name, val=pub_freq, err=e2
        ))

    # Rate of publisher
    pub_hdlr = rospy.Rate(pub_freq)
    # Publisher
    pub_obj = rospy.Publisher("pub_topic", String, queue_size=100)
    seq_num = 1
    # Main publishing loop
    while not rospy.is_shutdown():
        # Create a message object
        msg_obj = String()
        msg_obj.data = str(seq_num)
        pub_obj.publish(msg_obj)
        seq_num += 1
        # Log it
        rospy.logdebug("Published message [topic: {tname}] '{msgc}'".format(
            tname=pub_obj.name, msgc=msg_obj.data
        ))
        rospy.loginfo("Published '{msgc}'".format(
            msgc=msg_obj.data
        ))
        # Delay for adjusting to the pub_freq
        pub_hdlr.sleep()


# Entry point
if __name__ == '__main__':
    try:
        main()
    except rospy.ROSException as e:
        rospy.logfatal("ROS Exception: {0}".format(e))
